README.md
[![version](https://img.shields.io/github/v/release/lolouk44/xiaomi_mi_scale)](https://github.com/lolouk44/xiaomi_mi_scale/releases)
[![docker_badge](https://img.shields.io/docker/pulls/lolouk44/xiaomi-mi-scale)](https://hub.docker.com/r/lolouk44/xiaomi-mi-scale)
# Xiaomi Mi Scale

Code to read weight measurements from Xiaomi Body Scales.

## BREAKING CHANGE:
Please note that as off 0.2.0, the config is now located in options.json and no longer in the docker-compose / environment
Please read on for for more information.
This change was necessary to allow for unlimited number of users.

## Supported Scales:
Name | Model | Picture
--- | --- | :---:
[Mi Smart Scale 2](https://www.mi.com/global/scale) &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; &nbsp; | XMTZCO1HM, XMTZC04HM | ![Mi Scale_2](Screenshots/Mi_Smart_Scale_2_Thumb.png)
[Mi Body Composition Scale](https://www.mi.com/global/mi-body-composition-scale/) | XMTZC02HM | ![Mi Scale](Screenshots/Mi_Body_Composition_Scale_Thumb.png)
[Mi Body Composition Scale 2](https://c.mi.com/thread-2289389-1-0.html) | XMTZC05HM | ![Mi Body Composition Scale 2](Screenshots/Mi_Body_Composition_Scale_2_Thumb.png)


## Home Assistant Add-On:
If using Home Assistant (formerly known as hass.io), try instead the [Xiaomi Mi Scale Add-On for Home Assistant](https://github.com/lolouk44/hassio-addons/tree/master/mi-scale) based on this repository.

## Getting the Mac Address of your Scale:

1. Retrieve the scale's MAC Address from the Xiaomi Mi Fit App:

![MAC Address](Screenshots/MAC_Address.png)

## Setup & Configuration:
### Running script with Docker:

1. Supported platforms:
	1. linux/386
	1. linux/amd64
	1. linux/arm32v6
	1. linux/arm32v7
	1. linux/arm64v8
1. Open `docker-compose.yml` (see below) and edit the environment to suit your configuration...
1. Stand up the container - `docker-compose up -d`

### docker-compose:
```yaml
version: '3'
services:

  mi-scale:
    image: lolouk44/xiaomi-mi-scale:latest
    container_name: mi-scale
    restart: always

    network_mode: host
    privileged: true
    volumes:
      - ./data:/data
      - /var/run/dbus/:/var/run/dbus/:ro #needed for bleak
```
### options.json:
All the config needs to be in a file named `options.json`. You can get a copy of one with minimum config [here](./options.json)

List of options

Option | Type | Required | Description
--- | --- | --- | ---
MISCALE_MAC | string | Yes | Mac address of your scale
MQTT_HOST | string | Yes | MQTT Server (defaults to 127.0.0.1)
HCI_DEV | string | No | Bluetooth hci device to use. Defaults to hci0
MQTT_PREFIX | string | No | MQTT Topic Prefix. Defaults to miscale
MQTT_USERNAME | string | No | Username for MQTT server (comment out if not required)
MQTT_PASSWORD | string | No | Password for MQTT (comment out if not required)
MQTT_PORT | int | No | Defaults to 1883
MQTT_DISCOVERY | bool | No | MQTT Discovery for Home Assistant Defaults to true
MQTT_DISCOVERY_PREFIX | string | No | MQTT Discovery Prefix for Home Assistant. Defaults to homeassistant
MQTT_TLS_CACERTS | string | No | MQTT TLS connection: directory with CA certificate(s) that signed MQTT Server's TLS certificate, defaults to None (= no TLS connection)
MQTT_TLS_INSECURE | bool | No | MQTT TLS connection: don't verify hostname in TLS certificate, defaults to None (= always check hostname)
BLUEPY_PASSIVE_SCAN | bool | No | Try to set to true if getting an error like `Bluetooth connection error: Failed to execute management command ‘le on’` on a Raspberry Pi. Defaults to false
DEBUG_LEVEL | string | No | Logging level. Possible values: 'CRITICAL', 'ERROR', 'WARNING', 'INFO', 'DEBUG'. Defaults to 'INFO'.
USERS | List | Yes | List of users to add

Auto-gender selection/config -- This is used to create the calculations such as BMI, Water/Bone Mass etc...
Here is the logic used to assign a measured weight to a user:
- If the weight is within the range of a user's defined values for GT and LT, then it will be assigned (published) to that user.
- If the weight matches two separate user ranges, it will only be assigned to the first user that matched (so don't overlap ranges!)

User Option | Type | Required | Description
--- | --- | --- | ---
GT | int | Yes | Greater Than - Weight must be greater than this value - this will be the lower limit for the weight range of this user
LT | int | Yes | Less Than - Weight must be less than this value - this will be the upper limit for the weight range of this user
SEX | string | Yes | male / female
NAME | string | Yes | Name of the user
HEIGHT | int | Yes | Height (in cm) of the user
DOB | string | Yes | DOB (in yyyy-mm-dd format)

Note: The weight definitions must be in the same unit as the scale (kg, Lbs, jin)

### Running script directly on your host system:

***Note: Python 3.6 or higher is required to run the script manually***

***Note: this is now deprecated. It would still work provided the path to options.json is manually set in the code manually set path at line 39: `with open('/data/options.json') as json_file`***

1. Install python requirements (pip3 install -r requirements.txt)
1. Open `wrapper.sh` and configure your environment variables to suit your setup.
1. Add a cron-tab entry to wrapper like so:

```sh
@reboot bash /path/to/wrapper.sh
```

**NOTE**: Although once started the script runs continuously, it may take a few seconds for the data to be retrieved, computed and sent via mqtt.

## Home-Assistant Setup:
Under the `sensor` block, enter as many blocks as users configured in your environment variables.

```yaml
mqtt:
  sensor:
    - name: "Example Name Weight"
      state_topic: "miscale/USER_NAME/weight"
      value_template: "{{ value_json['weight'] }}"
      unit_of_measurement: "kg"
      json_attributes_topic: "miscale/USER_NAME/weight"
      icon: mdi:scale-bathroom
      # Below lines only needed if long term statistics are required
      state_class: "measurement"

    - name: "Example Name BMI"
      state_topic: "miscale/USER_NAME/weight"
      value_template: "{{ value_json['bmi'] }}"
      icon: mdi:human-pregnant
      unit_of_measurement: "kg/m2"
      # Below lines only needed if long term statistics are required
     state_class: "measurement"
```

![Mi Scale](Screenshots/HA_Lovelace_Card.png)

![Mi Scale](Screenshots/HA_Lovelace_Card_Details.png)

## Acknowledgements:
Thanks to @syssi (https://gist.github.com/syssi/4108a54877406dc231d95514e538bde9) and @prototux (https://github.com/wiecosystem/Bluetooth) for their initial code

Special thanks to [@ned-kelly](https://github.com/ned-kelly) for his help turning a "simple" python script into a fully fledged docker container

Thanks to [@bpaulin](https://github.com/bpaulin), [@AiiR42](https://github.com/AiiR42), [@andreasbrett](https://github.com/andreasbrett) for their PRs and collaboration


CHANGELOG.md
## [0.3.5] - 2022-10-10
### Added
- Added extra logging. Logging Level can be set from config file.
- Deprecated options MISCALE_VERSION, TIME_INTERVAL.
### Changed
- Restored HCI Settings

## [0.3.4] - 2022-10-05
### Changed
- Restored MQTT Discovery. ([fixes #65](https://github.com/lolouk44/hassio-addons/issues/65))
- Removed no longer needed MISCALE_VERSION

## [0.3.3] - 2022-10-03
### Changed
- Restoring handling of V1 scales. ([fixes #64](https://github.com/lolouk44/hassio-addons/issues/64))

## [0.3.0] - 2022-10-02
### Changed
- Stopped using deprecated/no longer supported bluepy library and replaced with bleak, requiring major code overhaul. ([fixes #59](https://github.com/lolouk44/hassio-addons/issues/59))
- Updated documentation to reflect MQTT integration (moved out of sensor config)
### Breaking Changes
- `/var/run/dbus/:/var/run/dbus/:ro` is needed in docker volumes for bleak to work
- If using a MiScale V1, make sure you add the MISCALE_VERSION to your options.json file (see doc)

## [0.2.4] - 2022-02-03
### Changed
- Changed time format for datestamp to contain timezone ([fixes #59](https://github.com/lolouk44/hassio-addons/issues/59))

## [0.2.3] - 2022-01-13
### Added
- Added support for Long Term Statistics (HA 2021.9 minimum required)
- Impedance posted to MQTT ([fixes #56](https://github.com/lolouk44/hassio-addons/issues/56))

## [0.2.2] - 2021-06-28
### Changed
- Fixed handling of MQTT_PORT and TIME_INTERVAL

## [0.2.1] - 2021-05-13
### Changed
- Fixed user lookup by non kg weight

## [0.2.0] - 2021-03-23
### Breaking Changes
Please note that as off 0.2.0, the config is now located in options.json and no longer in the docker-compose / environment. Be sure to check the [Doc](https://github.com/lolouk44/xiaomi_mi_scale/blob/master/README.md) for more information. This change was necessary to allow for unlimited number of users.
### Changed
- Unlimited users ! This unfortunately comes with a breaking change. Be sure to check the [Doc](https://github.com/lolouk44/xiaomi_mi_scale/blob/master/README.md) for more information
### Added
- Added BLUEPY_PASSIVE_SCAN Option for Raspberry Pi users experiencing Bluetooth troubles (in case you're not using the [add-on](https://github.com/lolouk44/hassio-addons/tree/master/mi-scale))

## [0.1.16] - 2021-03-16
### Changed
- Fixed MQTT Retain defaults

## [0.1.15] - 2021-03-12
### Changed
- Added MQTT Retain Option ([fixes #44](https://github.com/lolouk44/xiaomi_mi_scale/issues/44))
- Added MQTT TLS Option ([PR 43](https://github.com/lolouk44/xiaomi_mi_scale/pull/43))

## [0.1.14] - 2020-11-26
### Changed
- Reduced docker image size

## [0.1.13] - 2020-11-26
### Changed
- Fixed MQTT Discovery Message

## [0.1.12] - 2020-11-23
### Changed
- Updated workflow to automatically build docker images on new releases with version and latest tags

## [0.1.11] - 2020-11-23
### Changed
- Remove additional logging for Scale V1 that was used for testing
- Changed timestamp to default python format (fixes https://github.com/lolouk44/xiaomi_mi_scale/issues/29)
- Removed hard-coded 'unit_f_measurement' in the MQTT Discovery (fixes https://github.com/lolouk44/hassio-addons/issues/22)
- Fixed hard coded MQTT Discovery Prefix (fixes https://github.com/lolouk44/xiaomi_mi_scale/issues/35)
- Change measures format to be numbers instead of string where applicable (https://github.com/lolouk44/xiaomi_mi_scale/pull/36)
### Added
- Created workflow to automatically build docker images on new releases (Thanks [@AiiR42](https://github.com/AiiR42) for your help)


## [0.1.10] - 2020-09-09
### Changed
- Fixed issue with detection of boolean in MQTT_DISCOVERY (https://github.com/lolouk44/hassio-addons/issues/16 and https://github.com/lolouk44/xiaomi_mi_scale/issues/31)

## [0.1.9] - 2020-09-08
### Changed
- Fixed typo in MQTT message following the **breaking change** to snake_case attributes in 0.1.8

## [0.1.8] - 2020-09-08
### Breaking Changes
- Attributes are now snake_case (fixes https://github.com/lolouk44/xiaomi_mi_scale/issues/24)
### Changed
- Fixed default MQTT Prefix in config.json typo (fixes https://github.com/lolouk44/hassio-addons/issues/6)
- Fixed MQTT Discovery value check to discover
- Changed timestamp to default python format
- Changes the bluetooth reset from reset to down-wait-up (fixes https://github.com/lolouk44/hassio-addons/issues/13)
- Fixed hard coded hci0 to provided hci interface when performing a reset
- Fixed weight in Lbs not detected on Scale V1 (XMTZCO1HM) (fixes https://github.com/lolouk44/xiaomi_mi_scale/issues/28)
- Fixed body calculations for non kg weights
- Updated README
### Added
- Added unit to attributes

## [0.1.7] - 2020-07-06
### Added
- repository.json to make it a real add-on repo (fixes https://github.com/lolouk44/hassio-addons/issues/4)
## Changed
- Now truly handles optional config entries(fixes https://github.com/lolouk44/hassio-addons/issues/3)
- MQTT Discovery set wtih retain flag (fixes https://github.com/lolouk44/hassio-addons/issues/2)
- README updated to use Xiaomi Mi Fit App to retrieve the MAC Address (fixes https://github.com/lolouk44/xiaomi_mi_scale/pull/25)

## [0.1.6] - 2020-07-01
### Added
- Docker Image so install is quicker (no local build required).

## [0.1.5] - 2020-07-01
### Added
- MQTT Discovery Support.

## [0.1.4] - 2020-06-29
### Added
- First release (version in line with non Add-On script for ease of maintenance).


test_xiaomi_scale.py
#!/usr/bin/python3
from __future__ import print_function
import argparse
import binascii
import os
import sys
from bluepy import btle

MISCALE_MAC = 'XX:XX:XX:XX:XX:XX'
MISCALE_MAC = 'd4:ab:c8:b9:49:40'

if os.getenv('C', '1') == '0':
    ANSI_RED = ''
    ANSI_GREEN = ''
    ANSI_YELLOW = ''
    ANSI_CYAN = ''
    ANSI_WHITE = ''
    ANSI_OFF = ''
else:
    ANSI_CSI = "\033["
    ANSI_RED = ANSI_CSI + '31m'
    ANSI_GREEN = ANSI_CSI + '32m'
    ANSI_YELLOW = ANSI_CSI + '33m'
    ANSI_CYAN = ANSI_CSI + '36m'
    ANSI_WHITE = ANSI_CSI + '37m'
    ANSI_OFF = ANSI_CSI + '0m'


class ScanProcessor():

    def handleDiscovery(self, dev, isNewDev, isNewData):
        if dev.addr == MISCALE_MAC.lower() and isNewDev:
            print ('    Device: %s (%s), %d dBm %s. ' %
                   (
                       ANSI_WHITE + dev.addr + ANSI_OFF,
                       dev.addrType,
                       dev.rssi,
                       ('' if dev.connectable else '(not connectable)'))
                   , end='')
            for (sdid, desc, data) in dev.getScanData():
                if sdid == 22:
                   print('')
                   print ('data:')
                   print (data)

            if not dev.scanData:
                print ('\t(no data)')
            print




def main():

    scanner = btle.Scanner().withDelegate(ScanProcessor())

    print (ANSI_RED + "Scanning for devices..." + ANSI_OFF)
    devices = scanner.scan(5)

if __name__ == "__main__":
    main()

src/wrapper.sh
#!/bin/bash

sleep 60                      # Give the system time after a reboot to connect to WiFi before continuing
export MISCALE_MAC=00:00:00:00:00:00 # Mac address of your scale
export HCI_DEV=hci0                  # Bluetooth hci device to use
export MQTT_HOST=127.0.0.1           # MQTT Server (defaults to 127.0.0.1)
export MQTT_PREFIX=miscale           # MQTT Topic Prefix. Defaults to miscale
export MQTT_USERNAME=                # Username for MQTT server (comment out if not required)
export MQTT_PASSWORD=                # Password for MQTT (comment out if not required)
export MQTT_PORT=1883                # Defaults to 1883
export TIME_INTERVAL=30              # Time in sec between each query to the scale, to allow other applications to use the Bluetooth module. Defaults to 30
export MQTT_DISCOVERY=true           # Home Assistant Discovery (true/false), defaults to true
export MQTT_DISCOVERY_PREFIX=        # Home Assistant Discovery Prefix, defaults to homeassistant

export USER1_GT=70            # If the weight is greater than this number, we'll assume that we're weighing User #1
export USER1_SEX=male
export USER1_NAME=Jo          # Name of the user
export USER1_HEIGHT=175       # Height (in cm) of the user
export USER1_DOB=1990-01-01   # DOB (in yyyy-mm-dd format)

export USER2_LT=35            # If the weight is less than this number, we'll assume that we're weighing User #2
export USER2_SEX=female
export USER2_NAME=Sarah       # Name of the user
export USER2_HEIGHT=95        # Height (in cm) of the user
export USER2_DOB=1990-01-01   # DOB (in yyyy-mm-dd format)

export USER3_SEX=female
export USER3_NAME=Missy       # Name of the user
export USER3_HEIGHT=150       # Height (in cm) of the user
export USER3_DOB=1990-01-01   # DOB (in yyyy-mm-dd format)

MY_PWD="$( cd "$( dirname "${BASH_SOURCE[0]}" )" >/dev/null 2>&1 && pwd )"
python3 $MY_PWD/Xiaomi_Scale.py

src/Xiaomi_Scale.py
#!/usr/bin/python
# -*- coding: utf-8 -*-
import asyncio
import binascii
from bleak import BleakScanner
from collections import namedtuple
from datetime import datetime
import functools
import json
import paho.mqtt.publish as publish
import subprocess
import sys
import logging
import os

import Xiaomi_Scale_Body_Metrics

DEFAULT_DEBUG_LEVEL = "INFO"
VERSION = "0.3.5"



# User Config
class USER:
    def __init__(self, name, gt, lt, sex, height, dob):
        self.NAME, self.GT, self.LT, self.SEX, self.HEIGHT, self.DOB


def customUserDecoder(userDict):
    return namedtuple('USER', userDict.keys())(*userDict.values())

def MQTT_discovery():
    """Published MQTT Discovery information if enabled in options.json"""
    for MQTTUser in (USERS):
        message = '{"name": "' + MQTTUser.NAME + ' Weight",'
        message+= '"state_topic": "' + MQTT_PREFIX + '/' + MQTTUser.NAME + '/weight",'
        message+= '"value_template": "{{ value_json.weight }}",'
        message+= '"json_attributes_topic": "' + MQTT_PREFIX + '/' + MQTTUser.NAME + '/weight",'
        message+= '"icon": "mdi:scale-bathroom",'
        message+= '"state_class": "measurement"}'
        publish.single(
                        MQTT_DISCOVERY_PREFIX + '/sensor/' + MQTT_PREFIX + '/' + MQTTUser.NAME + '/config',
                        message,
                        retain=True,
                        hostname=MQTT_HOST,
                        port=MQTT_PORT,
                        auth={'username':MQTT_USERNAME, 'password':MQTT_PASSWORD},
                        tls=MQTT_TLS
                    )
    logging.info(f"MQTT Discovery Setup Completed...")

def check_weight(user, weight):
    return weight > user.GT and weight < user.LT

def GetAge(d1):
    d1 = datetime.strptime(d1, "%Y-%m-%d")
    d2 = datetime.strptime(datetime.today().strftime('%Y-%m-%d'),'%Y-%m-%d')
    return abs((d2 - d1).days)/365
    
def MQTT_publish(weight, unit, mitdatetime, hasImpedance, miimpedance):
    """Publishes weight data for the selected user"""
    if unit == "lbs": calcweight = round(weight * 0.4536, 2)
    if unit == "jin": calcweight = round(weight * 0.5, 2)
    if unit == "kg": calcweight = weight
    matcheduser = None
    for user in USERS:
        if(check_weight(user,weight)):
            matcheduser = user
            break
    if matcheduser is None:
        return
    height = matcheduser.HEIGHT
    age = GetAge(matcheduser.DOB)
    sex = matcheduser.SEX.lower()
    name = matcheduser.NAME

    lib = Xiaomi_Scale_Body_Metrics.bodyMetrics(calcweight, height, age, sex, 0)
    message = '{'
    message += '"weight":' + "{:.2f}".format(weight)
    message += ',"weight_unit":"' + str(unit) + '"'
    message += ',"bmi":' + "{:.2f}".format(lib.getBMI())
    message += ',"basal_metabolism":' + "{:.2f}".format(lib.getBMR())
    message += ',"visceral_fat":' + "{:.2f}".format(lib.getVisceralFat())

    if hasImpedance:
        lib = Xiaomi_Scale_Body_Metrics.bodyMetrics(calcweight, height, age, sex, int(miimpedance))
        bodyscale = ['Obese', 'Overweight', 'Thick-set', 'Lack-exerscise', 'Balanced', 'Balanced-muscular', 'Skinny', 'Balanced-skinny', 'Skinny-muscular']
        message += ',"lean_body_mass":' + "{:.2f}".format(lib.getLBMCoefficient())
        message += ',"body_fat":' + "{:.2f}".format(lib.getFatPercentage())
        message += ',"water":' + "{:.2f}".format(lib.getWaterPercentage())
        message += ',"bone_mass":' + "{:.2f}".format(lib.getBoneMass())
        message += ',"muscle_mass":' + "{:.2f}".format(lib.getMuscleMass())
        message += ',"protein":' + "{:.2f}".format(lib.getProteinPercentage())
        message += ',"body_type":"' + str(bodyscale[lib.getBodyType()]) + '"'
        message += ',"metabolic_age":' + "{:.0f}".format(lib.getMetabolicAge())
        message += ',"impedance":' + "{:.0f}".format(int(miimpedance))

    message += ',"timestamp":"' + mitdatetime + '"'
    message += '}'
    try:
        logging.info(f"Publishing data to topic {MQTT_PREFIX + '/' + name + '/weight'}: {message}")
        publish.single(
            MQTT_PREFIX + '/' + name + '/weight',
            message,
            retain=MQTT_RETAIN,
            hostname=MQTT_HOST,
            port=MQTT_PORT,
            auth={'username':MQTT_USERNAME, 'password':MQTT_PASSWORD},
            tls=MQTT_TLS
        )
        logging.info(f"Data Published ...")
    except Exception as error:
        logging.error(f"Could not publish to MQTT: {error}")
        raise

os.system('clear')

# Configuraiton...
# Trying To Load Config From options.json (HA Add-On)
try:
    with open('/data/options.json') as json_file:
        data = json.load(json_file)["options"]
        try:
            DEBUG_LEVEL = data["DEBUG_LEVEL"]
            if DEBUG_LEVEL not in ('CRITICAL','ERROR','WARNING','INFO','DEBUG','NOTSET'):
                DEBUG_LEVEL = DEFAULT_DEBUG_LEVEL
                logging.basicConfig(format='%(asctime)s - (%(levelname)s) %(message)s', level=DEBUG_LEVEL, datefmt='%Y-%m-%d %H:%M:%S')
                logging.info(f"-------------------------------------")
                logging.info(f"Starting Xiaomi mi Scale v{VERSION}...")
                logging.info(f"Loading Config From Options.json...")
                logging.warning(f"Invalid logging level provided, defaulting to {DEBUG_LEVEL}...")
            else:
                logging.basicConfig(format='%(asctime)s - (%(levelname)s) %(message)s', level=DEBUG_LEVEL, datefmt='%Y-%m-%d %H:%M:%S')
                logging.info(f"-------------------------------------")
                logging.info(f"Starting Xiaomi mi Scale v{VERSION}...")
                logging.info(f"Loading Config From Options.json...")
                logging.info(f"Logging Level Set to {DEBUG_LEVEL}...")
            # Prevent bleak log flooding
            bleak_logger = logging.getLogger("bleak")
            bleak_logger.setLevel(logging.INFO)
        except:
            DEBUG_LEVEL = DEFAULT_DEBUG_LEVEL
            logging.basicConfig(format='%(asctime)s - (%(levelname)s) %(message)s', level=DEBUG_LEVEL, datefmt='%Y-%m-%d %H:%M:%S')
            logging.info(f"-------------------------------------")
            logging.info(f"Starting Xiaomi mi Scale v{VERSION}...")
            logging.info(f"Loading Config From Options.json...")
            logging.info(f"No Logging Level Provided, Defaulting to  {DEBUG_LEVEL}...")
            # Prevent bleak log flooding
            bleak_logger = logging.getLogger("bleak")
            bleak_logger.setLevel(logging.INFO)
            pass
        try:
            MISCALE_MAC = data["MISCALE_MAC"]
            logging.debug(f"MISCALE_MAC read from config: {MISCALE_MAC}")

        except:
            logging.error(f"MAC Address not provided...")
            raise
        try:
            MISCALE_VERSION = data["MISCALE_VERSION"]
            logging.info(f"MISCALE_VERSION option is deprecated and can safely be removed from config...")
        except:
            pass
        try:
            MQTT_USERNAME = data["MQTT_USERNAME"]
            logging.debug(f"MQTT_USERNAME read from config: {MQTT_USERNAME}")
        except:
            MQTT_USERNAME = "username"
            logging.debug(f"MQTT_USERNAME defaulted to: {MQTT_USERNAME}")
            pass
        try:
            MQTT_PASSWORD = data["MQTT_PASSWORD"]
            logging.debug(f"MQTT_PASSWORD read from config: ***")
        except:
            MQTT_PASSWORD = None
            logging.debug(f"MQTT_PASSWORD defaulted to: {MQTT_PASSWORD}")
            pass
        try:
            MQTT_HOST = data["MQTT_HOST"]
            logging.debug(f"MQTT_HOST read from config: {MQTT_HOST}")
        except:
            logging.error(f"MQTT Host not provided...")
            raise
        try:
            MQTT_RETAIN = data["MQTT_RETAIN"]
            logging.debug(f"MQTT_RETAIN read from config: {MQTT_RETAIN}")
        except:
            MQTT_RETAIN = True
            logging.debug(f"MQTT_RETAIN defaulted to: {MQTT_RETAIN}")
            pass
        try:
            MQTT_PORT = data["MQTT_PORT"]
            logging.debug(f"MQTT_PORT read from config: {MQTT_PORT}")
            if(type(MQTT_PORT) != int):
                logging.warning(f"Converting MQTT_PORT to integer...")
                MQTT_PORT = int(MQTT_PORT)
        except:
            MQTT_PORT = 1883
            logging.debug(f"MQTT_PORT defaulted to: {MQTT_PORT}")
            pass
        try:
            MQTT_TLS_CACERTS = data["MQTT_TLS_CACERTS"]
            logging.debug(f"MQTT_TLS_CACERTS read from config: {MQTT_TLS_CACERTS}")
        except:
            MQTT_TLS_CACERTS = None
            logging.debug(f"MQTT_TLS_CACERTS defaulted to: {MQTT_TLS_CACERTS}")
            pass
        try:
            MQTT_TLS_INSECURE = data["MQTT_TLS_INSECURE"]
            logging.debug(f"MQTT_TLS_INSECURE read from config: {MQTT_TLS_INSECURE}")
        except:
            MQTT_TLS_INSECURE = None
            logging.debug(f"MQTT_TLS_INSECURE defaulted to: {MQTT_TLS_INSECURE}")
            pass
        try:
            MQTT_PREFIX = data["MQTT_PREFIX"]
            logging.debug(f"MQTT_PREFIX read from config: {MQTT_PREFIX}")
        except:
            MQTT_PREFIX = "miscale"
            logging.debug(f"MQTT_PREFIX defaulted to: {MQTT_PREFIX}")
            pass
        try:
            TIME_INTERVAL = data["TIME_INTERVAL"]
            logging.info(f"TIME_INTERVAL option is deprecated and can safely be removed from config...")
        except:
            pass
        try:
            MQTT_DISCOVERY = data["MQTT_DISCOVERY"]
            logging.debug(f"MQTT_DISCOVERY read from config: {MQTT_DISCOVERY}")
        except:
            MQTT_DISCOVERY = True
            logging.debug(f"MQTT_DISCOVERY defaulted to: {MQTT_DISCOVERY}")
            pass
        try:
            MQTT_DISCOVERY_PREFIX = data["MQTT_DISCOVERY_PREFIX"]
            logging.debug(f"MQTT_DISCOVERY_PREFIX read from config: {MQTT_DISCOVERY_PREFIX}")
        except:
            if MQTT_DISCOVERY:
                logging.warning(f"MQTT Discovery enabled but no MQTT Prefix provided, defaulting to 'homeassistant'...")
                MQTT_DISCOVERY_PREFIX = "homeassistant"
            pass
        try:
            HCI_DEV = data["HCI_DEV"].lower()
            logging.debug(f"HCI_DEV read from config: {HCI_DEV}")
        except:
            HCI_DEV = "hci0"
            logging.debug(f"HCI_DEV defaulted to: {HCI_DEV}")
            pass
        try:
            BLUEPY_PASSIVE_SCAN = data["BLUEPY_PASSIVE_SCAN"]
            logging.debug(f"BLUEPY_PASSIVE_SCAN read from config: {BLUEPY_PASSIVE_SCAN}")
        except:
            BLUEPY_PASSIVE_SCAN = False
            logging.debug(f"BLUEPY_PASSIVE_SCAN defaulted to: {BLUEPY_PASSIVE_SCAN}")
            pass

        if MQTT_TLS_CACERTS in [None, '', 'Path to CA Cert File']:
            MQTT_TLS = None
        else:
            MQTT_TLS = {'ca_certs':MQTT_TLS_CACERTS, 'insecure':MQTT_TLS_INSECURE}

        USERS = []
        for user in data["USERS"]:    
            try:
                user = json.dumps(user)
                user = json.loads(user, object_hook=customUserDecoder)
                if user.GT > user.LT:
                    raise ValueError("GT can not be larger than LT - user {user.Name}")  
                USERS.append(user)
            except:
                logging.error(f"{sys.exc_info()[1]}")
                raise
        OLD_MEASURE = None
        logging.info(f"Config Loaded...")

# Failed to open options.json
except FileNotFoundError as error:
    DEBUG_LEVEL = DEFAULT_DEBUG_LEVEL
    logging.basicConfig(format='%(asctime)s - (%(levelname)s) %(message)s', level=DEBUG_LEVEL, datefmt='%Y-%m-%d %H:%M:%S')
    logging.info(f"-------------------------------------")
    logging.info(f"Starting Xiaomi mi Scale v{VERSION}...")
    logging.info(f"Loading Config From Options.json...")
    logging.error(f"options.json file missing... {error}")
    # Prevent bleak log flooding
    bleak_logger = logging.getLogger("bleak")
    bleak_logger.setLevel(logging.INFO)
    raise


async def main(MISCALE_MAC):
    stop_event = asyncio.Event()

    # TODO: add something that calls stop_event.set()

    def callback(device, advertising_data):
        global OLD_MEASURE
        if device.address.lower() == MISCALE_MAC:
            logging.debug(f"miscale found, with advertising_data: {advertising_data}")
            try:
                ### Xiaomi V2 Scale ###
                data = binascii.b2a_hex(advertising_data.service_data['0000181b-0000-1000-8000-00805f9b34fb']).decode('ascii')
                logging.debug(f"miscale v2 found (service data: 0000181b-0000-1000-8000-00805f9b34fb)")
                data = "1b18" + data # Remnant from previous code. Needs to be cleaned in the future
                data2 = bytes.fromhex(data[4:])
                ctrlByte1 = data2[1]
                isStabilized = ctrlByte1 & (1<<5)
                hasImpedance = ctrlByte1 & (1<<1)
                measunit = data[4:6]
                measured = int((data[28:30] + data[26:28]), 16) * 0.01
                unit = ''
                if measunit == "03": unit = 'lbs'
                if measunit == "02": unit = 'kg' ; measured = measured / 2
                miimpedance = str(int((data[24:26] + data[22:24]), 16))
                if unit and isStabilized:
                    if OLD_MEASURE != round(measured, 2) + int(miimpedance):
                        OLD_MEASURE = round(measured, 2) + int(miimpedance)
                        MQTT_publish(round(measured, 2), unit, str(datetime.now().strftime('%Y-%m-%dT%H:%M:%S+00:00')), hasImpedance, miimpedance)
            except:
                pass
            try:
                ### Xiaomi V1 Scale ###
                data = binascii.b2a_hex(advertising_data.service_data['0000181d-0000-1000-8000-00805f9b34fb']).decode('ascii')
                logging.debug(f"miscale v1 found (service data: 0000181d-0000-1000-8000-00805f9b34fb)")
                data = "1d18" + data # Remnant from previous code. Needs to be cleaned in the future
                measunit = data[4:6]
                measured = int((data[8:10] + data[6:8]), 16) * 0.01
                unit = ''
                if measunit.startswith(('03', 'a3')): unit = 'lbs'
                if measunit.startswith(('12', 'b2')): unit = 'jin'
                if measunit.startswith(('22', 'a2')): unit = 'kg' ; measured = measured / 2
                if unit:
                    if OLD_MEASURE != round(measured, 2):
                        OLD_MEASURE = round(measured, 2)
                        MQTT_publish(round(measured, 2), unit, str(datetime.now().strftime('%Y-%m-%dT%H:%M:%S+00:00')), "", "")
            except:
                pass
        pass

    async with BleakScanner(
        callback,
        device=f"{HCI_DEV}"
        ) as scanner:
        ...
        # Important! Wait for an event to trigger stop, otherwise scanner
        # will stop immediately.
        await stop_event.wait()

        
if __name__ == "__main__":
    if MQTT_DISCOVERY:
        MQTT_discovery()
    logging.info(f"-------------------------------------")
    logging.info(f"Initialization Completed, Waiting for Scale...")
    try:
        asyncio.run(main(MISCALE_MAC.lower()))
    except Exception as error:
        logging.error(f"Unable to connect to Bluetooth: {error}")
        pass